Rebase #2431 against proposed/3.0.0

alexwlchan opened the pull request:
This is really quite unexciting: #2431 had a tentative +1 from @Lukasa, pending a rebase and retargeting against 3.0.0, but it’s been sat unloved for nearly eight months. 😢 

This patch does the required rebase and targets the 3.0.0 branch.

Lukasa commented on tests/test_requests.py:
```diff
@@ -1203,6 +1203,46 @@ def test_request_and_response_are_pickleable(self, httpbin):
+        r.iter_content = mock_iter_content
+
+        assert list(r.iter_lines(delimiter='\r\n')) == \
+            ''.join(mock_chunks).split('\r\n')
```
I'd like these asserts cleaned up if at all possible, they're kinda complex. Let's break some stuff up so the assert is somewhat self-describing.

Lukasa commented on requests/models.py:
```diff
@@ -753,23 +753,36 @@ def iter_lines(self, chunk_size=ITER_CHUNK_SIZE, decode_unicode=None, delimiter=
+            # item to prepend to the next chunk. However, only do this for
+            # non-delimiter mode if the chunk does not match the end of the
+            # last line.
+            if delimiter or (lines[-1] and lines[-1][-1] == chunk[-1]):
```
Can we break out the parenthetical here and give it a name that hopefully describes what we're actually checking?

ianjepperson commented:
Woah, zombie PR back from the dead!  I haven't worked in Python for over a year now, but I'll see if I can address the suggestions this weekend.

alexwlchan commented:
I had a go at addressing the review markups, but I feel like more testing is required.

I feel like there’s enough subtle behaviour here that we’d be better off with a whole stack of tests (probably using `@pytest.mark.parametrize`) that tests a whole variety of cases here. Which I may or may not get around to writing sometime this week. At a minimum, the cases described in [this comment](https://github.com/kennethreitz/requests/pull/2431#issuecomment-72333964).

sigmavirus24 commented on requests/models.py:
```diff
@@ -737,42 +737,70 @@ def generate():
 
         return chunks
 
+    import q
```
What is `q`?

alexwlchan replied:
Oh damn, somebody checked in his debugging code without checking the diff, didn't he?

My oopsie, will fix.

([q](https://pypi.python.org/pypi/q/2.6) is a module for quick and dirty debugging that has no business in a production codebase)

Lukasa commented:
Yeah, I think more testing here would be fab. Not sure how we're going to get it though, especially if @alexwlchan doesn't have time to do it.

alexwlchan commented:
I’ve stuck a note in my todo list to come back and write more tests here. I’ve built up a bit of a backlog of unfinished PRs, but hopefully I’ll find time to get back to this over Christmas.

kennethreitz commented:
Closing this out due to lack of activity. Feel free to re-open once you get a chance!
